### PR TITLE
Completed Excel tutorial

### PR DESCRIPTION
**Pull request by AlexJerabek**
|        Q        |                    A                    |
| --------------- | --------------------------------------- |
| Bug fix?        | no                            |
| New feature?    | no                           |
| New sample?     | yes                               |
| Related issues? |  |

## What's in this Pull Request?

This PR adds a completed version of the [Excel tutorial from Office Add-ins documentation](https://learn.microsoft.com/en-us/office/dev/add-ins/tutorials/excel-tutorial).

### CODE REVIEW

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/dialogs/popup.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+(function () {
+    "use strict";
+
+    Office.onReady()
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/My Office Add-in/src/dialogs/popup.js:4]: Again, tutorial should use consistency and this should match the same pattern as Office.onReady in taskpane.js.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+ * See LICENSE in the project root for license information.
+ */
+
+/* global console, document, Excel, Office */
+
+Office.onReady((info) => {
+  if (info.host === Office.HostType.Excel) {
+    // Determine if the user's version of Office supports all the Office.js APIs that are used in the tutorial.
+    if (!Office.context.requirements.isSetSupported('ExcelApi', '1.7')) {
+      console.log('Sorry. The tutorial add-in uses Excel.js APIs that are not available in your version of Office.');
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:12]: tutorial probably shouldn't enable the buttons if the APIs are unavailable.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      ["1/1/2017", "The Phone Company", "Communications", "120"],
+      ["1/2/2017", "Northwind Electric Cars", "Transportation", "142.33"],
+      ["1/5/2017", "Best For You Organics Company", "Groceries", "27.9"],
+      ["1/10/2017", "Coho Vineyard", "Restaurant", "33"],
+      ["1/11/2017", "Bellows College", "Education", "350.1"],
+      ["1/15/2017", "Trey Research", "Other", "135"],
+      ["1/15/2017", "Best For You Organics Company", "Groceries", "97.88"]
+    ]);
+
+    expensesTable.columns.getItemAt(3).getRange().numberFormat = [['\u20AC#,##0.00']];
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:47]: would love a link to more docs on that number format.

File: Samples/tutorials/excel-tutorial/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Excel Tutorial - Completed
+
+This sample is the result of completing the [Tutorial: Create an Excel task pane add-in](https://learn.microsoft.com/office/dev/add-ins/tutorials/excel-tutorial). It was constructed with the [Yeoman generator for Office Add-ins](https://learn.microsoft.com/office/dev/add-ins/develop/yeoman-generator-overview).
+
+The tutorial gives step-by-step instructions on how to add functionality alongside explanations as to why code is being added. This sample is best used if you've completed the tutorial and want to start from a more complete project. It's also helpful when debugging problems you encounter during the tutorial.
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/README.md:5]:
```suggestion
The tutorial gives step-by-step instructions on how to add functionality alongside explanations as to why code is being added. Use this sample if you want to explore and try the completed code, or if you need to debug any issues you may have encountered while following the tutorial.
```

Comment (davidchesnut) [reply Samples/tutorials/excel-tutorial/README.md:5]: Just avoiding suggestion that you should do the tutorial first.

File: Samples/tutorials/excel-tutorial/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## Run the add-in
+
+1. If the local web server is already running and your add-in is already loaded in Excel, proceed to step 2. Otherwise, start the local web server and sideload your add-in:
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/README.md:9]: I'm not following how you could already have the web server running. I think we should drop this step.

File: Samples/tutorials/excel-tutorial/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+- To test your add-in in Excel, run the following command in the root directory of your project. This starts the local web server (if it's not already running) and opens Excel with your add-in loaded.
+
+  - Windows: `npm start`
+  - Mac: `npm run dev-server`
+
+- To test your add-in in Excel on the web, run the following command in the root directory of your project. When you run this command, the local web server starts. Replace "{url}" with the URL of an Excel document on your OneDrive or a SharePoint library to which you have permissions.
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/README.md:16]: Need to say to use the My Office add-in folder (not the root just above it).

**Review by davidchesnut — CHANGES_REQUESTED** [review]
Nice! A few comments provided.

- We need jsdoc comments on all the tutorial code added. 
- Would be nice to avoid mixing await/async with promise patterns like .next. I've been told this can get confusing for developers. But this may require updating the article as well.

File: Samples/tutorials/excel-tutorial/My Office Add-in/.gitignore [hunk 7]
```diff
@@ -0,0 +1,2 @@
+.vscode/
```
[the remark this answers is not in the document]
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/My Office Add-in/.gitignore]: Cool. We should add that to the contribution guidance and make sure other samples follow it.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/commands/commands.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ */
+
+/* global global, Office, self, window */
+
+Office.onReady(() => {
+  // If needed, Office.js is ready to be called
+});
+
+/**
+ * Shows a notification when the add-in command is executed.
```
[the remark this answers is not in the document]
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/My Office Add-in/src/commands/commands.js:13]: Let me tag @millerds here for awareness.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/commands/commands.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    if (sheet.protection.protected) {
+      sheet.protection.unprotect();
+    } else {
+      sheet.protection.protect();
+    }
+
+    await context.sync();
+  })
+    .catch(function (error) {
```
[the remark this answers is not in the document]
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/My Office Add-in/src/commands/commands.js:61]: Do you want that to be part of this PR? Happy to do it, but it'll take more time.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js (hunk 2 above)
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:12]: TBH, I think the tutorial shouldn't include this `isSetSupported` stuff. If you're targeting something this old, you have a very specific audience in mind. Thoughts?

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  })
+    .catch(function (error) {
+      console.log("Error: " + error);
+      if (error instanceof OfficeExtension.Error) {
+        console.log("Debug info: " + JSON.stringify(error.debugInfo));
+      }
+    });
+}
+
+async function freezeHeader() {
```
[the remark this answers is not in the document]
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:128]: Hmm... this was working fine for me. I'll try to repro the issue.

File: Samples/tutorials/excel-tutorial/README.md (hunk 4 above)
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/README.md:5]: But don't we want people to take the tutorial first? I think if someone had this PNP as their entry point, we'd want them to go back and follow the step-by-step instructions, unless they really don't want to.

**Review by AlexJerabek — COMMENTED** [review]
Thanks @davidchesnut. Updated.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (davidchesnut) [reply Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:117]: Never mind.

[the remark this answers is not in the document]
Comment (AlexJerabek) [reply Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:112]: FWIW, this happens when the chart is selected. Focus matters.

File: Samples/tutorials/excel-tutorial/README.md (hunk 4 above)
Comment (davidchesnut) [reply Samples/tutorials/excel-tutorial/README.md:23]: I'm assuming there's different learning styles and some developers may want to just run the code and step through or study it before doing all the copy/paste work from the tutorial.

File: Samples/tutorials/excel-tutorial/README.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+  - `npm run start:web -- --document https://contoso.sharepoint.com/:t:/g/EZGxP7ksiE5DuxvY638G798BpuhwluxCMfF1WZQj3VYhYQ?e=F4QM1R`
+  - `npm run start:web -- --document https://1drv.ms/x/s!jkcH7spkM4EGgcZUgqthk4IK3NOypVw?e=Z6G1qp`
+  - `npm run start:web -- --document https://contoso-my.sharepoint-df.com/:t:/p/user/EQda453DNTpFnl1bFPhOVR0BwlrzetbXvnaRYii2lDr_oQ?e=RSccmNP`
+
+  > NOTE: If you are developing on a Mac, enclose the {url} in single quotation marks. Do not do this on Windows.
+
+    If your add-in doesn't sideload in the document, manually sideload it by following the instructions in Manually sideload add-ins to Office on the web.
+
+1. If the add-in task pane isn't already open in Excel, go to the Home tab and choose the **Show Taskpane** button in the ribbon to open it.
```
Comment (davidchesnut) [inline Samples/tutorials/excel-tutorial/README.md:84]: The previous sections cause this numbering to restart. Probably should indent the previous sections.

**Review by davidchesnut — APPROVED** [review]
Just fix the readme line numbering (see comment) otherwise looks great!

**Comment by AlexJerabek** [comment]
Adding @ElizabethSamuel-MSFT for a review just to double check everything runs.

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+      ["1/5/2017", "Best For You Organics Company", "Groceries", "27.9"],
+      ["1/10/2017", "Coho Vineyard", "Restaurant", "33"],
+      ["1/11/2017", "Bellows College", "Education", "350.1"],
+      ["1/15/2017", "Trey Research", "Other", "135"],
+      ["1/15/2017", "Best For You Organics Company", "Groceries", "97.88"]
+    ]);
+
+    
+    // Learn more about the Excel number format syntax in this article:
+    // https://support.microsoft.com/office/number-format-codes-5026bbd6-04bc-48cd-bf33-80f18b4eae68
```
Comment (ElizabethSamuel-MSFT) [inline Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:47]:
```suggestion
    // https://support.microsoft.com/office/5026bbd6-04bc-48cd-bf33-80f18b4eae68
```

File: Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    chart.legend.format.fill.setSolidColor("white");
+    chart.dataLabels.format.font.size = 15;
+    chart.dataLabels.format.font.color = "black";
+    chart.series.getItemAt(0).name = 'Value in \u20AC';
+
+    await context.sync();
+  });
+}
+
+async function freezeHeader() {
```
Comment (ElizabethSamuel-MSFT) [inline Samples/tutorials/excel-tutorial/My Office Add-in/src/taskpane/taskpane.js:116]: add a function description?

**Comment by AlexJerabek** [comment]
Thanks for the feedback @ElizabethSamuel-MSFT. I've updated the PR accordingly (as well as the corresponding PR in the Docs repo).

File: README.md [hunk 14]
```diff
@@ -11,6 +11,14 @@ The following samples show how to build the simplest Office Add-in with only a m
 * [PowerPoint "Hello world" add-in](Samples/hello-world/powerpoint-hello-world)
 * [Word "Hello world" add-in](Samples/hello-world/word-hello-world)
 
+### Completed tutorials
+
+The following sample is the completed version of the [Excel add-in tutorial](https://learn.microsoft.com/office/dev/add-ins/tutorials/excel-tutorial).
+
+| Name           | Description  |
+| -------------- | ------------ |
+| [Tutorial: Create an Excel task pane add-in](https://learn.microsoft.com/office/dev/add-ins/tutorials/excel-tutorial) | This sample shows how to create an Excel add-in with a task pane and command ribbon buttons. The add-in shows how to create and sort a table, make a chart, freeze rows, protect a worksheet, and pop-up a dialog box. |
```
Comment (ElizabethSamuel-MSFT) [inline README.md:20]: consider "display" or "pop up"
```suggestion
| [Tutorial: Create an Excel task pane add-in](https://learn.microsoft.com/office/dev/add-ins/tutorials/excel-tutorial) | This sample shows how to create an Excel add-in with a task pane and command ribbon buttons. The add-in shows how to create and sort a table, make a chart, freeze a row, protect a worksheet, and display a dialog box. |
```

**Review by ElizabethSamuel-MSFT — APPROVED** [review]

**Pull request merged**